mavros_extras/src/plugins/fake_gps.cpp

```diff
@@ -0,0 +1,309 @@
+	Eigen::Vector3d old_ecef;	//!< previous geocentric position [m]
+	double old_stamp;		//!< previous stamp [s]
+
+	inline Eigen::Vector3d enu_to_ecef_transform(const Eigen::Affine3d &map_point)
+	{
+		/**
+		 * @brief Apply transform from ENU to ECEF:
+		 *
+		 * ϕ = x
+		 * λ = y
+		 *
+		 * The rotation is composed by a clockwise rotation over the east-axis
+		 * by an angle of π - ϕ followed by a clockwise over the Z-axis by
+		 * an angle of π + λ
+		 *
+		 * R = [-sinλ       -cosλ*sinϕ   cosλ*cosϕ
+		 *       cosϕ       -sinλ*sinϕ   sinλ*cosϕ
+		 *       0.0         cosϕ        sinϕ     ]
+		 *
+		 * x, y, z = R * [∂x, ∂y, ∂z]
+		 */
+		Eigen::Matrix3d R;
+
+		const double sin_x = std::sin(map_point.translation().x());
+		const double sin_y = std::sin(map_point.translation().y());
+		const double cos_x = std::cos(map_point.translation().x());
+		const double cos_y = std::cos(map_point.translation().z());
+
+		R << -sin_y, -cos_y * sin_x, cos_y * cos_x,
+		cos_x, -sin_y * sin_x, sin_y * cos_x,
+		0.0,    cos_x,         sin_x;
+
+		return R * map_point.translation();
+	}
+
+	/* -*- mid-level helpers and low-level send -*- */
+
+	/**
+	 * @brief Send fake GPS coordinates through HIL_GPS Mavlink msg
+	 */
+	void send_fake_gps(const ros::Time &stamp, const Eigen::Vector3d &ecef_offset) {
+		// Throttle incoming messages to 5hz
+		if ((ros::Time::now() - last_pos_time) < gps_period) {
+			return;
+		}
+		last_pos_time = ros::Time::now();
+
+		/**
+		 * @note: HIL_GPS messages are accepted on PX4 Firmware out of HIL mode,
+		 * if use_hil_gps flag is set (param MAV_USEHILGPS = 1).
+		 * @todo: add GPS_INPUT msg as an alternative, as Ardupilot already supports it
+		 */
+		mavlink::common::msg::HIL_GPS fix {};
+
+		Eigen::Vector3d geodetic;
+		Eigen::Vector3d current_ecef(ecef_origin.x() + ecef_offset.x(),
+					ecef_origin.y() + ecef_offset.y(),
+					ecef_origin.z() + ecef_offset.z());
+
+		try {
+			earth.Reverse(current_ecef.x(), current_ecef.y(), current_ecef.z(),
+						geodetic.x(), geodetic.y(), geodetic.z());
+		}
+		catch (const std::exception& e) {
+			ROS_INFO_STREAM("FGPS: Caught exception: " << e.what() << std::endl);
+		}
+
+		Eigen::Vector3d vel = ((old_ecef - current_ecef) / (stamp.toSec() - old_stamp)) * 1e2;// [cm/s]
+
+		// compute course over ground
+		double cog;
+		if (vel.x() == 0 && vel.y() == 0) {
+			cog = 0;
+		}
+		else if (vel.x() >= 0 && vel.y() < 0) {
+			cog = M_PI * 5 / 2 - atan2(vel.x(), vel.y());
+		}
+		else {
+			cog = M_PI / 2 - atan2(vel.x(), vel.y());
+		}
+
+		// Fill in and send message
+		fix.time_usec = stamp.toNSec() / 1000;	// [useconds]
+		fix.lat = geodetic.x() * 1e7;		// [degrees * 1e7]
+		fix.lon = geodetic.y() * 1e7;		// [degrees * 1e7]
+		fix.alt = (geodetic.z() + GeographicLib::Geoid::ELLIPSOIDTOGEOID *
+					(*m_uas->egm96_5)(geodetic.x(), geodetic.y())) * 1e3;	// [meters * 1e3]
+		fix.vel = sqrt(vel.x() * vel.x() + vel.y() * vel.y());				// [cm/s]
```

Comment:
Yeah I suppose that works but at a computational level, would that really be more efficient?